chainer/functions/normalization/batch_normalization.py

```diff
@@ -152,205 +395,69 @@ def forward(self, inputs):
-                if isinstance(self.running_var, intel64.ideep.mdarray):
-                    self.running_var.inplace_axpby(
-                        self.decay, (1 - self.decay), self.var * adjust)
-                else:
-                    self.running_var *= self.decay
-                    self.running_var += self.var * adjust * (1 - self.decay)
-
-        elif self.use_cudnn:
-            if self.running_mean is not None:
-                mean = self.running_mean
-                var = self.running_var
-            else:
-                # Create dummies.
-                mean = xp.zeros_like(gamma, dtype=x.dtype)
-                var = xp.zeros_like(gamma, dtype=x.dtype)
-
-            # self.mean and self.inv_std are used as buffers to save
-            # intermediate results computed during forward pass. These buffers
-            # are used to speed-up backward pass.
-            y, self.mean, self.inv_std = (
-                cudnn.batch_normalization_forward_training(
-                    x, gamma, beta, mean, var, None, None,
-                    self.eps, self.decay, self.mode.is_for_conv2d,
-                    self.mode.get_cudnn_mode(), chainer.is_debug()))
-        else:
-            # Generic CPU and GPU implementation
-
-            interm_dtype = numpy.promote_types(x.dtype, gamma.dtype)
-
-            gamma = gamma[expander].astype(interm_dtype, copy=False)
-            beta = beta[expander].astype(interm_dtype, copy=False)
-            self.mean = x.mean(axis=self.axis, dtype=interm_dtype)
-            var = x.var(axis=self.axis, dtype=interm_dtype)
-            if xp is numpy:
-                self.inv_std = numpy.reciprocal(numpy.sqrt(
-                    var + self.eps, dtype=interm_dtype))
-            else:
-                self.inv_std = cuda.cupyx.rsqrt(
-                    var + self.eps, dtype=interm_dtype)
 
-            y = _apply_bn_fwd(xp, x, self.mean[expander],
-                              self.inv_std[expander], gamma, beta)
-
-            # Update running statistics if given
-            if self.running_mean is not None:
-                m = x.size // gamma.size
-                adjust = m / max(m - 1., 1.)  # unbiased estimation
-
-                xp = backend.get_array_module(
-                    self.running_mean, self.running_var)
-
-                if xp is chainerx:
-                    self.running_mean, self.running_var = backend.from_chx(
-                        (self.running_mean, self.running_var))
-
-                self.running_mean *= self.decay
-                self.running_mean += (1 - self.decay) * self.mean
-                self.running_var *= self.decay
-                self.running_var += (1 - self.decay) * adjust * var
-
-                if xp is chainerx:
-                    self.running_mean = backend.to_chx(self.running_mean)
-                    self.running_var = backend.to_chx(self.running_var)
+        self.mode = _BNMode(x, gamma, self.key_axis)
```

Comment:
I think it's not necessary to keep `self.mode` for backward.

Reply:
addressed in 8be6e49